include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,19 +121,210 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+
+
+        constexpr ::uint32_t __elems_per_wg = _ElemsPerItem * _WGSize;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc = sycl::accessor<_ValueType, 1, sycl::access_mode::read_write, sycl::target::local>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<__par_backend_hetero::__tunable_kernel_name<_GroupScanKernel, __elems_per_wg, _IsFullGroup, _Inclusive>>(sycl::nd_range<1>(_WGSize, _WGSize), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                const auto& __subgroup = __self_item.get_sub_group();
+                const auto __item_id = __self_item.get_local_linear_id();
+                int __subgroup_id = __subgroup.get_group_id();
+                int __id_in_subgroup = __subgroup.get_local_id();
+                int __subgroup_size = __subgroup.get_local_linear_range();
+
+                if constexpr (_IsFullGroup)
+                {
+                    #pragma unroll
+                    for (uint16_t __i = 0; __i < _ElemsPerItem; ++__i)
+                    {
+                       auto __idx = __i * _WGSize + __subgroup_id * __subgroup_size;
+                       auto __val = __unary_op(__subgroup.load(__in.begin() + __idx));
+                       __subgroup.store(__lacc.get_pointer() + __idx, __val);
+                    }
+                }
+                else
+                {
+                    #pragma unroll
+                    for (uint16_t __i = 0; __i < _ElemsPerItem; ++__i)
+                    {
+                       auto __idx = __i * _WGSize + __item_id;
+                       auto __val = __idx < __n ? __in[__idx] : _ValueType{};
+                       __lacc[__idx] = __val;
+                    }
+
+                }
+
+                __group_scan<_ValueType>(__group, __lacc.get_pointer(), __lacc.get_pointer() + __n, __bin_op, __init);
+
+                if constexpr (_IsFullGroup)
+                {
+                    #pragma unroll
+                    for (uint16_t __i = 0; __i < _ElemsPerItem; ++__i)
+                    {
+                       auto __idx = __i * _WGSize + __subgroup_id * __subgroup_size;
+                       auto __val = __subgroup.load(__lacc.get_pointer() + __idx);
+                       __subgroup.store(__out.begin() + __idx, __val);
+                    }
+                }
+                else
+                {
+                    #pragma unroll
+                    for (uint16_t __i = 0; __i < _ElemsPerItem; ++__i)
+                    {
+                       auto __idx = __i * _WGSize + __item_id;
+                       if (__idx < __n)
+                           __out[__idx] = __lacc[__idx];
+                    }
```

Comment:
I'm not sure it saves us much, but we can change the loop bounds based on `__n`, rather than branching internally to the loop.